src/Console/ConsoleOptionParser.php

```diff
@@ -581,10 +581,15 @@ public function parse(array $argv, ?ConsoleIo $io = null): array
         }
 
         foreach ($this->_args as $i => $arg) {
-            if ($arg->isRequired() && !isset($args[$i])) {
-                throw new ConsoleException(
-                    sprintf('Missing required argument. The `%s` argument is required.', $arg->name())
-                );
+            if (!isset($args[$i])) {
```

Comment:
How will adding arguments with default values work if those arguments are added after arguments without default values?

Reply:
Exactly like the positional arguments of PHP functions.

Reply:
Sorry, I asked the inverse scenario of what I was actually trying to express.

```php
$parser->addArgument('optional', ['default' => 'a'])
  ->addArgument('required')
```

This feels like a mistake that we could and should try to prevent much like PHP does.

Reply:
We could throw an exception here

Reply:
With this code
```php
    public function buildOptionParser(ConsoleOptionParser $parser): ConsoleOptionParser
    {
        $parser = parent::buildOptionParser($parser);
        $parser->addArgument('optional', ['default' => 'a'])->addArgument('required');

        return $parser;
    }

    public function execute(Arguments $args, ConsoleIo $io)
    {
        debug($args->getArguments());
    }
```
```bash
/home/user$ cake Example -h
Usage:
cake example [-h] [-q] [-v] [<optional>] [<required>]

Options:

--help, -h     Display this help.
--quiet, -q    Enable quiet output.
--verbose, -v  Enable verbose output.

Arguments:

optional   (optional) default: "a"
required   (optional)

/home/user$ cake Example 
APP/Command\ExampleCommand.php (line 48)
########## DEBUG ##########
[
  (int) 0 => 'a'
]
###########################

/home/user$ cake Example b
APP/Command\ExampleCommand.php (line 48)
########## DEBUG ##########
[
  (int) 0 => 'b'
]
###########################

/home/user$ cake Example b c
APP/Command\ExampleCommand.php (line 48)
########## DEBUG ##########
[
  (int) 0 => 'b',
  (int) 1 => 'c'
]
###########################
```

Same behavior as in `function example(optional='a', required=null)`

Reply:
> We could throw an exception here

The "required" behavior overrides the "default" behavior.

Reply:
Required before optional arguments:
```php
    public function buildOptionParser(ConsoleOptionParser $parser): ConsoleOptionParser
    {
        $parser = parent::buildOptionParser($parser);
        $parser
               ->addArgument('required', ['required' => true])
               ->addArgument('optional', ['default' => 'a']);

        return $parser;
    }
```
If required is after optional argument Exception is thrown:
```php
    public function buildOptionParser(ConsoleOptionParser $parser): ConsoleOptionParser
    {
        $parser = parent::buildOptionParser($parser);
        $parser
               ->addArgument('optional', ['default' => 'a'])
               ->addArgument('required', ['required' => true]);

        return $parser;
    }
```
The exception is raised whether a default value is specified or not. An optional positional argument simply cannot come before a required positional argument.
This is not new behavior or a modification of existing behavior.